py_mini_racer/extension/mini_racer_extension.cc

```diff
@@ -208,23 +212,37 @@ static void init_v8() {
     platform_lock.unlock();
 }
 
-static void breaker(std::timed_mutex& breaker_mutex, void * d) {
-  EvalParams* data = (EvalParams*)d;
 
-  if (!breaker_mutex.try_lock_for(std::chrono::milliseconds(data->timeout))) {
-    data->result->timed_out = true;
-    data->context_info->isolate->TerminateExecution();
-  }
-}
+class BreakerTask: public Task {
+    public:
+    BreakerTask(std::shared_ptr<EvalParams> params): params_(params), deadline_(current_platform->MonotonicallyIncreasingTime() + (params->timeout / 1000.0)) {
+    }
+    void Run() {
+      std::shared_ptr<EvalParams> params = params_.lock();
+      double remaining = deadline_ - current_platform->MonotonicallyIncreasingTime();
+
+      // Do not run the breaker if too late (params do not exist anymore)
+      if (params && remaining > 0) {
+          if (!params->breaker_mutex.try_lock_for(std::chrono::duration<double>(remaining))) {
```

Comment:
This seems to have a race. I'm assuming the shared_ptr `params` can be `reset` somewhere in the code, hence the check on line 225. But on this line it could have been reset already.